packages/react-reconciler/src/ReactFiberCommitWork.js

```diff
@@ -295,8 +295,21 @@ function commitHookEffectList(
       if ((effect.tag & mountTag) !== NoHookEffect) {
         // Mount
         const create = effect.create;
-        const destroy = create();
-        effect.destroy = typeof destroy === 'function' ? destroy : null;
+        let destroy = create();
+        if (typeof destroy !== 'function') {
+          if (__DEV__ && destroy !== null && destroy !== undefined) {
+            warningWithoutStack(
+              false,
+              'useEffect function must return a cleanup function or nothing.%s%s',
+              typeof destroy.then === 'function'
+                ? ' Promises and async functions are not supported.'
```

Comment:
Maybe add "but you can call an async function without waiting for it" or something. Or I expect we'll keep getting issues asking for this.